### PR TITLE
Only allow valid return types for willReturn

### LINKED ISSUE #3122: Prevent runtime type error due to wrong return value configuration
`willReturn()` should raise an error when a return value is configured that does not match the stubbed method's return value.

```php
<?php declare(strict_types=1);

class C
{
    public function m(): int
    {
        return 1;
    }
}

final class Test extends PHPUnit\Framework\TestCase
{
    public function testOne(): void
    {
        $o = $this->createMock(C::class);
        
        $o->method('m')->willReturn('not an integer');
        
        $o->m();
    }
}
```
### Actual: `TypeError` when stubbed method is invoked
```
PHPUnit 7.1-gfafd3c555 by Sebastian Bergmann and contributors.

E                                                                   1 / 1 (100%)

Time: 68 ms, Memory: 4.00MB

There was 1 error:

1) Test::testOne
TypeError: Return value of Mock_C_06141907::m() must be of the type integer, string returned

/home/sb/Test.php:19

ERRORS!
Tests: 1, Assertions: 0, Errors: 1.
```

### PR DESCRIPTION
**Pull request by MichelHartmann**
Another enhancement from the code sprint at @FLYERALARM in Würzburg:

This fixes: #3122 and #3602

If you would like me to change something before it can merged just give me a hint.

### CODE REVIEW

File: src/Framework/MockObject/ConfigurableMethod.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+    }
+
+    public function getName(): string
+    {
+        return $this->name;
+    }
+
+    public function mayReturn($value): bool
+    {
+        if (isNull($value) && $this->returnType->allowsNull()) {
```
Comment (psalm-shepherd) [inline src/Framework/MockObject/ConfigurableMethod.php]: Too many arguments for method isNull - expecting 0 but saw 1
```
        if (isNull($value) && $this->returnType->allowsNull()) {
            ^^^^^^^^^^^^^^
```

File: src/Framework/MockObject/InvocationMocker.php [hunk 2]
```diff
@@ -93,7 +93,7 @@ public function expects(MatcherInvocation $matcher): BuilderInvocationMocker
         return new BuilderInvocationMocker(
             $this,
             $matcher,
-            $this->configurableMethods
+            ...$this->configurableMethods
```
Comment (psalm-shepherd) [inline src/Framework/MockObject/InvocationMocker.php]: Argument 3 of PHPUnit\Framework\MockObject\Builder\InvocationMocker::__construct expects PHPUnit\Framework\MockObject\ConfigurableMethod, string provided
```
            ...$this->configurableMethods
               ^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: src/Framework/MockObject/MockMethod.php [hunk 3]
```diff
@@ -124,7 +118,7 @@ public static function fromReflection(\ReflectionMethod $method, bool $callOrigi
             $modifier,
             self::getMethodParameters($method),
             self::getMethodParameters($method, true),
-            $returnType,
+            self::deriveReturnType($method),
```
Comment (psalm-shepherd) [inline src/Framework/MockObject/MockMethod.php]: Argument 7 of PHPUnit\Framework\MockObject\MockMethod::__construct cannot be null, possibly null value provided
```
            self::deriveReturnType($method),
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
```

File: src/Framework/MockObject/Type.php [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+            case 'object':
+                return new ObjectType(TypeName::fromQualifiedName(\get_class($value)), $allowsNull);
+            default:
+                return self::fromName($type, $allowsNull);
+        }
+    }
+
+    public static function fromName(?string $typeName, bool $allowsNull): self
+    {
+        switch (\mb_strtolower($typeName)) {
```
Comment (psalm-shepherd) [inline src/Framework/MockObject/Type.php]: Argument 1 of mb_strtolower cannot be null, possibly null value provided
```
        switch (\mb_strtolower($typeName)) {
                               ^^^^^^^^^
```

File: src/Framework/MockObject/TypeName.php [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+    public function getSimpleName(): string
+    {
+        return $this->simpleName;
+    }
+
+    public function getQualifiedName(): string
+    {
+        return $this->isNamespaced()
+             ? $this->namespaceName . '\\' . $this->simpleName
```
Comment (psalm-shepherd) [inline src/Framework/MockObject/TypeName.php]: Cannot concatenate with a possibly null null|string
```
             ? $this->namespaceName . '\\' . $this->simpleName
               ^^^^^^^^^^^^^^^^^^^^
```

**Comment by psalm-shepherd** [comment]
Psalm found some errors

**Comment by psalm-shepherd** [comment]
Psalm didn’t find any errors!

**Comment by sebastianbergmann** [comment]
Thank you, @MichelHartmann. This is a lot to review, I hope I'll get to it soon.

**Comment by sebastianbergmann** [comment]
@MichelHartmann Sorry that it took so long to merge this. Thanks again!

**Pull request merged**